Multistage Docker Builds

mmlb opened the pull request:
## Description

Changes the osie container to use multistage builds for every binary built from source.

## Why is this needed

Make builds faster. Fixes #129

## How Has This Been Tested?

Builds run, all the tools are there and can be invoked.

mmlb committed:
ci: Drop the packet-networking exclude from the black check

packet-networking no longer lives in the repo.

Signed-off-by: Manuel Mendez <mmendez@equinix.com>

mmlb committed:
nix: Update nixpkgs revision

To get docker buildx support.

Signed-off-by: Manuel Mendez <mmendez@equinix.com>

mmlb committed:
docker: Yo dawg I heard you like layers

Now that we squash the final image, there's no need to worry
about avoiding all these layers. Now we get better caching
when something goes wrong halfway through a build.

Signed-off-by: Manuel Mendez <mmendez@equinix.com>

mmlb committed:
docker: Refactor source builds to account for squashing

No need to install/remove build-essential for each binary since
we now squash at the end.

Signed-off-by: Manuel Mendez <mmendez@equinix.com>

mmlb committed:
docker: Refactor how source builds are done

Use the Makefile's DESTDIR handling to install to an outdir
and then we can copy the files of interest into /. This is
mostly in prep for when we do multi-stage builds.

Signed-off-by: Manuel Mendez <mmendez@equinix.com>

mmlb committed:
docker: Multistage builds!

For better cachability! If one of the multistages needs to be rebuilt docker only
rebuilds it and after where the files are COPYd from. And if paired with buildx
all the images will be built in parallel.

Signed-off-by: Manuel Mendez <mmendez@equinix.com>

markjacksonfishing commented on docker/Dockerfile:
```diff
@@ -1,4 +1,34 @@
-FROM ubuntu:xenial
+FROM ubuntu:xenial as builder
```
Should we be pinning a version here?

mmlb replied:
I don't think its necessary. Doesn't buy us much. The apt caches are still blown away so we'd still need to do an `apt-get update` for most of the things we want. I've got a jira ticket to possibly explore pinning the version to a date and what not but that'll be left for another day.

markjacksonfishing commented on docker/Dockerfile:
```diff
@@ -1,4 +1,34 @@
-FROM ubuntu:xenial
+FROM ubuntu:xenial as builder
+RUN apt-get update -y && apt-get install -y build-essential curl
```
`--no-install-recommends` should also be in here

mmlb replied:
I'll give this a try.

mmlb replied:
Nope that doesn't work, and seeing as this is just the build image I'm ok with having some extra junk in there.

markjacksonfishing commented on docker/Dockerfile:
```diff
@@ -1,4 +1,34 @@
+        https://dl.dell.com/FOLDER05920767M/1/DellEMC-iDRACTools-Web-LX-9.4.0-3732_A00.tar.gz \
+        http://linux.dell.com/repo/community/openmanage/940/bionic/pool/main/s/srvadmin-omilcore/srvadmin-omilcore_9.4.0_amd64.deb
+RUN tar -xvf DellEMC-iDRACTools-Web-LX-9.4.0-3732_A00.tar.gz
+RUN alien iDRACTools/racadm/RHEL8/x86_64/*.rpm
```
Use ./*glob* or -- *glob* so names with dashes won't become options.

mmlb replied:
that doesn't apply hear since the `iDrac...` does the same effect as `./`

markjacksonfishing reviewed (changes requested):
I know this is still in draft but some initial comments

mmlb committed:
docker: Drop sources.list.aarch64

We can use docker buildx and the bootstrap container image to run a
vanilla aarch64 ubuntu image which has all the correct sources already.

Signed-off-by: Manuel Mendez <mmendez@equinix.com>

mmlb committed:
make: Use docker buildx to build the osie containers

This simplifies the Dockerfiles and building them. This
is also the current "standard" way to build multiplatform
images so lets stop swimming against the current.

Signed-off-by: Manuel Mendez <mmendez@equinix.com>

mmlb commented:
This is in the back burner for the moment, need to figure out how to squash the `buildx` built images.

nshalman commented:
@mmlb should we keep this open?

jacobweinstock commented:
Discussed this with @mmlb in the Triage party and he will re-open this if/when desired/needed.
